File: swift/EgyNames/Sources/EgyNames/RulesConfig.swift
import Foundation

/// Loader for the shared, cross-SDK rule config (`data/logic_config.json`,
/// synced into `Resources/logic_config.json` by `scripts/sync-catalog.sh`).
///
/// Mirrors `python/src/egy_names/_rules_config.py`. If the config file is
/// missing or malformed, falls back to the values last known correct from
/// the audits that produced it, so the library never hard-fails on a
/// packaging mistake.
struct RulesInferRule: Codable, Sendable {
    let script: String?
    let match: String?
    let prefix: RulesStringOrList?
    let suffix: RulesStringOrList?
    let contains: RulesStringOrList?
    let value: String?
    let confidence: Double?
}

/// Decodes a JSON field that may be either a single string or an array of strings.
enum RulesStringOrList: Codable, Sendable {
    case single(String)
    case list([String])

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let s = try? container.decode(String.self) {
            self = .single(s)
        } else {
            self = .list(try container.decode([String].self))
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .single(let s): try container.encode(s)
        case .list(let l): try container.encode(l)
        }
    }

    var values: [String] {
        switch self {
        case .single(let s): return [s]
        case .list(let l): return l
        }
    }
}

private struct RulesQualityConfig: Codable, Sendable {
    let nonPersonalAr: [String]?
    let uncertainMeaningMarkers: [String]?
    let lowConfidenceShareEpsilon: Double?
    let kunyaExemptPrefixes: [String]?

    enum CodingKeys: String, CodingKey {
        case nonPersonalAr = "non_personal_ar"
        case uncertainMeaningMarkers = "uncertain_meaning_markers"
        case lowConfidenceShareEpsilon = "low_confidence_share_epsilon"
        case kunyaExemptPrefixes = "kunya_exempt_prefixes"
    }
}

private struct RulesInferThresholds: Codable, Sendable {
    let genderMinP: Double?
    let muslimMinP: Double?
    let christianMinP: Double?
    let roleMinP: Double?

    enum CodingKeys: String, CodingKey {
        case genderMinP = "gender_min_p"
        case muslimMinP = "muslim_min_p"
        case christianMinP = "christian_min_p"
        case roleMinP = "role_min_p"
    }
}

private struct RulesInferRules: Codable, Sendable {
    let gender: [RulesInferRule]?
    let religion: [RulesInferRule]?
    let role: [RulesInferRule]?
}

private struct RulesConfigFile: Codable, Sendable {
    let quality: RulesQualityConfig?
    let inferThresholds: RulesInferThresholds?
    let inferRules: RulesInferRules?

    enum CodingKeys: String, CodingKey {
        case quality
        case inferThresholds = "infer_thresholds"
        case inferRules = "infer_rules"
    }
}

public struct InferThresholds: Sendable {
    public let genderMinP: Double
    public let muslimMinP: Double
    public let christianMinP: Double
    public let roleMinP: Double
}

public final class RulesConfig: @unchecked Sendable {
    private static let lock = NSLock()
    private static var loaded = false

    private static var _nonPersonalAr: Set<String> = []
    private static var _uncertainMeaningMarkers: [String] = []
    private static var _lowConfidenceShareEpsilon: Double = 0.0001
    private static var _kunyaExemptPrefixes: [String] = []
    private static var _inferThresholds = InferThresholds(
        genderMinP: 0.70, muslimMinP: 0.85, christianMinP: 0.90, roleMinP: 0.88
    )
    private static var _inferRulesGender: [RulesInferRule] = []
    private static var _inferRulesReligion: [RulesInferRule] = []
    private static var _inferRulesRole: [RulesInferRule] = []

    private static let fallbackNonPersonalAr: [String] = [
        "الله", "الرجل", "الرجال", "شربه", "لافندي", "لفندي", "ماء", "البيت",
    ]

    private static let fallbackUncertainMeaningMarkers: [String] = [
        "غير واضح", "لا يوجد معنى", "غير معروف",
        "قد يكون تحريف", "تحريفاً", "تحريفًا",
    ]

    private static let fallbackKunyaExemptPrefixes: [String] = ["أبو", "ابو", "أم", "ام"]

    private static func ensureLoaded() {
        if loaded { return }
        lock.lock()
        defer { lock.unlock() }
        if loaded { return }
        load()
        loaded = true
    }

    private static func load() {
        var config: RulesConfigFile?

        #if SWIFT_PACKAGE
        if let url = Bundle.module.url(forResource: "logic_config", withExtension: "json"),
           let data = try? Data(contentsOf: url) {
            config = try? JSONDecoder().decode(RulesConfigFile.self, from: data)
        }
        #endif

        guard let cfg = config else {
            applyFallback()
            return
        }

        _nonPersonalAr = Set(cfg.quality?.nonPersonalAr ?? fallbackNonPersonalAr)
        _uncertainMeaningMarkers = cfg.quality?.uncertainMeaningMarkers ?? fallbackUncertainMeaningMarkers
        _lowConfidenceShareEpsilon = cfg.quality?.lowConfidenceShareEpsilon ?? 0.0001
        _kunyaExemptPrefixes = cfg.quality?.kunyaExemptPrefixes ?? fallbackKunyaExemptPrefixes

        _inferThresholds = InferThresholds(
            genderMinP: cfg.inferThresholds?.genderMinP ?? 0.70,
            muslimMinP: cfg.inferThresholds?.muslimMinP ?? 0.85,
            christianMinP: cfg.inferThresholds?.christianMinP ?? 0.90,
            roleMinP: cfg.inferThresholds?.roleMinP ?? 0.88
        )

        _inferRulesGender = cfg.inferRules?.gender ?? []
        _inferRulesReligion = cfg.inferRules?.religion ?? []
        _inferRulesRole = cfg.inferRules?.role ?? []
    }

    private static func applyFallback() {
        _nonPersonalAr = Set(fallbackNonPersonalAr)
        _uncertainMeaningMarkers = fallbackUncertainMeaningMarkers
        _lowConfidenceShareEpsilon = 0.0001
        _kunyaExemptPrefixes = fallbackKunyaExemptPrefixes
        _inferThresholds = InferThresholds(
            genderMinP: 0.70, muslimMinP: 0.85, christianMinP: 0.90, roleMinP: 0.88
        )
        _inferRulesGender = []
        _inferRulesReligion = []
        _inferRulesRole = []
    }

    public static var nonPersonalAr: Set<String> {
        ensureLoaded()
        return _nonPersonalAr
    }

    public static var uncertainMeaningMarkers: [String] {
        ensureLoaded()
        return _uncertainMeaningMarkers
    }

    public static var lowConfidenceShareEpsilon: Double {
        ensureLoaded()
        return _lowConfidenceShareEpsilon
    }

    public static var kunyaExemptPrefixes: [String] {
        ensureLoaded()
        return _kunyaExemptPrefixes
    }

    public static var inferThresholds: InferThresholds {
        ensureLoaded()
        return _inferThresholds
    }

    static func inferRules(_ kind: String) -> [RulesInferRule] {
        ensureLoaded()
        switch kind {
        case "gender": return _inferRulesGender
        case "religion": return _inferRulesReligion
        case "role": return _inferRulesRole
        default: return []
        }
    }
}

File: java/egy-names/src/main/java/com/afify/egynames/index/RulesConfig.java
package com.afify.egynames.index;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.File;
import java.io.FileInputStream;
import java.io.InputStream;
import java.util.*;

/**
 * Loader for the shared, cross-SDK rule config ({@code logic_config.json}).
 *
 * <p>Mirrors {@code python/src/egy_names/_rules_config.py}: a single source of
 * truth for non-personal surfaces, low-confidence detection thresholds, and
 * the (currently unused by Java) ML-abstention thresholds/rule tables. If the
 * config file is missing or malformed, falls back to the values last known
 * correct, so the library never hard-fails on a packaging mistake.
 */
public final class RulesConfig {

    private static final Set<String> NON_PERSONAL_AR_FALLBACK = Set.of(
            "الله", "الرجل", "الرجال", "شربه", "لافندي", "لفندي", "ماء", "البيت"
    );

    private static final List<String> UNCERTAIN_MEANING_MARKERS_FALLBACK = List.of(
            "غير واضح", "لا يوجد معنى", "غير معروف",
            "قد يكون تحريف", "تحريفاً", "تحريفًا"
    );

    private static final double LOW_CONFIDENCE_SHARE_EPSILON_FALLBACK = 0.0001;

    private static final List<String> KUNYA_EXEMPT_PREFIXES_FALLBACK = List.of("أبو", "ابو", "أم", "ام");

    private static final Map<String, Double> INFER_THRESHOLDS_FALLBACK = Map.of(
            "gender_min_p", 0.70,
            "muslim_min_p", 0.85,
            "christian_min_p", 0.90,
            "role_min_p", 0.88
    );

    private static volatile Set<String> nonPersonalAr;
    private static volatile List<String> uncertainMeaningMarkers;
    private static volatile Double lowConfidenceShareEpsilon;
    private static volatile List<String> kunyaExemptPrefixes;
    private static volatile Map<String, Double> inferThresholds;
    private static volatile Map<String, List<JsonNode>> inferRules;

    private RulesConfig() {}

    private static synchronized JsonNode load() {
        try {
            InputStream is = RulesConfig.class.getClassLoader().getResourceAsStream("logic_config.json");
            if (is == null) {
                is = RulesConfig.class.getResourceAsStream("/logic_config.json");
            }
            if (is == null && new File("src/main/resources/logic_config.json").exists()) {
                is = new FileInputStream("src/main/resources/logic_config.json");
            }
            if (is == null) {
                return null;
            }
            try (InputStream stream = is) {
                ObjectMapper mapper = new ObjectMapper();
                return mapper.readTree(stream);
            }
        } catch (Exception e) {
            return null;
        }
    }

    private static List<String> asStringList(JsonNode node) {
        List<String> out = new ArrayList<>();
        if (node != null && node.isArray()) {
            for (JsonNode n : node) {
                out.add(n.asText());
            }
        }
        return out;
    }

    private static synchronized void ensureLoaded() {
        if (nonPersonalAr != null) return;

        JsonNode root = load();
        JsonNode quality = root != null ? root.get("quality") : null;

        if (quality != null && quality.has("non_personal_ar")) {
            nonPersonalAr = new HashSet<>(asStringList(quality.get("non_personal_ar")));
        } else {
            nonPersonalAr = new HashSet<>(NON_PERSONAL_AR_FALLBACK);
        }

        if (quality != null && quality.has("uncertain_meaning_markers")) {
            uncertainMeaningMarkers = asStringList(quality.get("uncertain_meaning_markers"));
        } else {
            uncertainMeaningMarkers = UNCERTAIN_MEANING_MARKERS_FALLBACK;
        }

        if (quality != null && quality.has("low_confidence_share_epsilon")) {
            lowConfidenceShareEpsilon = quality.get("low_confidence_share_epsilon").asDouble();
        } else {
            lowConfidenceShareEpsilon = LOW_CONFIDENCE_SHARE_EPSILON_FALLBACK;
        }

        if (quality != null && quality.has("kunya_exempt_prefixes")) {
            kunyaExemptPrefixes = asStringList(quality.get("kunya_exempt_prefixes"));
        } else {
            kunyaExemptPrefixes = KUNYA_EXEMPT_PREFIXES_FALLBACK;
        }

        JsonNode thresholdsNode = root != null ? root.get("infer_thresholds") : null;
        if (thresholdsNode != null) {
            Map<String, Double> m = new HashMap<>();
            Iterator<Map.Entry<String, JsonNode>> fields = thresholdsNode.fields();
            while (fields.hasNext()) {
                Map.Entry<String, JsonNode> f = fields.next();
                if (f.getValue().isNumber()) {
                    m.put(f.getKey(), f.getValue().asDouble());
                }
            }
            inferThresholds = m.isEmpty() ? INFER_THRESHOLDS_FALLBACK : m;
        } else {
            inferThresholds = INFER_THRESHOLDS_FALLBACK;
        }

        Map<String, List<JsonNode>> rules = new HashMap<>();
        JsonNode rulesNode = root != null ? root.get("infer_rules") : null;
        for (String kind : List.of("gender", "religion", "role")) {
            List<JsonNode> list = new ArrayList<>();
            if (rulesNode != null && rulesNode.has(kind) && rulesNode.get(kind).isArray()) {
                for (JsonNode n : rulesNode.get(kind)) {
                    list.add(n);
                }
            }
            rules.put(kind, list);
        }
        inferRules = rules;
    }

    public static Set<String> nonPersonalAr() {
        ensureLoaded();
        return nonPersonalAr;
    }

    public static List<String> uncertainMeaningMarkers() {
        ensureLoaded();
        return uncertainMeaningMarkers;
    }

    public static double lowConfidenceShareEpsilon() {
        ensureLoaded();
        return lowConfidenceShareEpsilon;
    }

    public static List<String> kunyaExemptPrefixes() {
        ensureLoaded();
        return kunyaExemptPrefixes;
    }

    public static Map<String, Double> inferThresholds() {
        ensureLoaded();
        return inferThresholds;
    }

    /** Rule table for 'gender' | 'religion' | 'role'. Loaded for forward compatibility only. */
    public static List<JsonNode> inferRules(String kind) {
        ensureLoaded();
        return inferRules.getOrDefault(kind, Collections.emptyList());
    }
}
